Sources/PhishingDetection/PhishingDetectionDataStore.swift

```diff
@@ -0,0 +1,134 @@
+        createFileDataStore()
+    }
+
+    func createFileDataStore() {
+        do {
+            let fileManager = FileManager.default
+            let appSupportDirectory = try fileManager.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
+            dataStore = appSupportDirectory.appendingPathComponent(Bundle.main.bundleIdentifier!, isDirectory: true)
+            try fileManager.createDirectory(at: dataStore!, withIntermediateDirectories: true, attributes: nil)
+        } catch {
+            os_log(.debug, log: .phishingDetection, "\(self): 🔴 Error creating phishing protection data directory: \(error)")
+        }
+    }
+
+    public func writeData() {
+        let encoder = JSONEncoder()
+        do {
+            let hashPrefixesData = try encoder.encode(Array(hashPrefixes))
+            let filterSetData = try encoder.encode(Array(filterSet))
+            let revision = try encoder.encode(self.currentRevision)
+
+            let hashPrefixesFileURL = dataStore!.appendingPathComponent(hashPrefixFilename)
+            let filterSetFileURL = dataStore!.appendingPathComponent(filterSetFilename)
+            let revisionFileURL = dataStore!.appendingPathComponent(revisionFilename)
+
+            try hashPrefixesData.write(to: hashPrefixesFileURL)
+            try filterSetData.write(to: filterSetFileURL)
+            try revision.write(to: revisionFileURL)
+        } catch {
+            os_log(.debug, log: .phishingDetection, "\(self): 🔴 Error saving phishing protection data: \(error)")
+        }
+    }
+
+    public func loadData() async {
+        let decoder = JSONDecoder()
+        do {
+            let hashPrefixesFileURL = dataStore!.appendingPathComponent("hashPrefixes.json")
+            let filterSetFileURL = dataStore!.appendingPathComponent("filterSet.json")
+            let revisionFileURL = dataStore!.appendingPathComponent("revision.txt")
+
+            let hashPrefixesData = try Data(contentsOf: hashPrefixesFileURL)
+            let filterSetData = try Data(contentsOf: filterSetFileURL)
+            let revisionData = try Data(contentsOf: revisionFileURL)
+
+            hashPrefixes = Set(try decoder.decode([String].self, from: hashPrefixesData))
+            filterSet = Set(try decoder.decode([Filter].self, from: filterSetData))
+            currentRevision = try decoder.decode(Int.self, from: revisionData)
+
+            if (hashPrefixes.isEmpty && filterSet.isEmpty) || currentRevision == 0 {
+                throw PhishingDetectionDataError.empty
+            }
+        } catch {
+            os_log(.debug, log: .phishingDetection, "\(self): 🔴 Error loading phishing protection data: \(error). Reloading from embedded dataset.")
+            self.currentRevision = dataProvider.embeddedRevision
+            self.hashPrefixes = dataProvider.loadEmbeddedHashPrefixes()
+            self.filterSet = dataProvider.loadEmbeddedFilterSet()
+        }
+    }
+}
```

Comment:
I think it would be nice to separate some responsabilities and have a FileStorageManager
`lass FileStorageManager {
    let dataStoreURL: URL

    init(dataStoreURL: URL) {
        self.dataStoreURL = dataStoreURL
        createDirectoryIfNeeded()
    }

    private func createDirectoryIfNeeded() {
        do {
            try FileManager.default.createDirectory(at: dataStoreURL, withIntermediateDirectories: true, attributes: nil)
        } catch {
            os_log(.error, log: .default, "Failed to create directory: %{public}@", error.localizedDescription)
        }
    }

    func write(data: Data, to filename: String) {
        let fileURL = dataStoreURL.appendingPathComponent(filename)
        do {
            try data.write(to: fileURL)
        } catch {
            os_log(.error, log: .default, "Error writing to %{public}@: %{public}@", filename, error.localizedDescription)
        }
    }

    func read(from filename: String) -> Data? {
        let fileURL = dataStoreURL.appendingPathComponent(filename)
        do {
            return try Data(contentsOf: fileURL)
        } catch {
            os_log(.error, log: .default, "Error reading from %{public}@: %{public}@", filename, error.localizedDescription)
            return nil
        }
    }
}`

something like this